File: apps/mobile/android/app/src/main/java/xyz/speakeasyapp/app/voicefilter/dsp/VoiceFilterDsp.kt
package xyz.speakeasyapp.app.voicefilter.dsp

import xyz.speakeasyapp.app.voicefilter.SampleFilter
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.pow

/**
 * The Phase 5j Private Call voice filter.
 *
 * v1 algorithm (locked in plan): pitch + formant shift by N semitones
 * via a tape-head granular shifter ([GranularPitchShifter]). Both
 * pitch and formants move together — that's enough for a recognizable
 * mask. Independent pitch / formant shifting is a v2 problem and can
 * land behind this same interface without touching PR-B2's WebRTC
 * fork.
 *
 * Wired to PR-B2's forked `SpeakeasyAudioRecord` via [SampleFilter];
 * see [xyz.speakeasyapp.app.voicefilter.VoiceFilterModule] for the
 * call-time toggle.
 *
 * Allocation-free per-frame: the temp short[]s are sized for the
 * largest plausible frame ([MAX_FRAME_SAMPLES]) and reused.
 */
class VoiceFilterDsp(
    /**
     * Pitch shift in semitones. Negative = down (sounds "deeper",
     * more disguised); positive = up. Default −2 matches the
     * locked plan choice for back-compat with pre-rc.17 callers.
     */
    semitones: Float = -2f,
    /**
     * Phase 2b: formant shift in semitones, INDEPENDENT of pitch.
     * 0 = preserve original formants (the vocal tract sounds the
     * same size as the speaker's, just with shifted pitch — like
     * helium without the chipmunk effect). Negative = formants
     * down (larger-sounding vocal tract, deeper resonance).
     * Positive = formants up (smaller, brighter, more "bright").
     * Default 0 means "match pitch shift" if the caller doesn't
     * specify — preserves rc.18 character for back-compat.
     *
     * Only honored by the phase-vocoder backend; granular ignores
     * (it can't separate pitch from formants).
     */
    formantSemitones: Float = 0f,
    /**
     * Per-frame latency budget in microseconds. Matches the 80ms p50
     * plan ceiling; we set the trip at 120ms (p95) so a noisy single
     * frame doesn't kill the call.
     */
    budgetMicros: Long = 120_000L,
    /** Injected for tests; production uses [System.nanoTime]. */
    private val nowNanos: () -> Long = { System.nanoTime() },
) : SampleFilter {

  private val factor: Float = 2.0.pow((semitones / 12.0)).toFloat()
  private val formantFactor: Float =
      2.0.pow((formantSemitones / 12.0)).toFloat()
  // Phase 2a: phase vocoder replaces the granular shifter as the
  // default. Lower latency (~10ms vs ~21ms) and no crackle, at the
  // cost of some pitch-shift artifacts on transients (plosives) and
  // sustained vowels (faint metallic edge). The boolean below lets
  // us flip back to granular fast if field testing finds the new
  // path worse on real hardware.
  // Phase 2b: vocoder takes formantFactor at init for independent
  // pitch/formant control. Granular ignores formant (can't do it).
  private val shifter: PitchShifter =
    if (USE_PHASE_VOCODER) PhaseVocoderShifterAdapter(formantFactor)
    else GranularShifterAdapter()
  private val guard = LatencyGuard(budgetMicros)

  /**
   * Scratch buffer for stereo→mono collapse and back-write. PCM16
   * samples (not bytes). Re-used across calls.
   */
  private val scratch = ShortArray(MAX_FRAME_SAMPLES)

  /** Most recently observed channel count; informational. */
  @Volatile var lastChannelCount: Int = 0
    private set

  override fun process(
      samples: ByteBuffer,
      sampleRateHz: Int,
      channelCount: Int,
  ): Boolean {
    if (guard.isTripped()) {
      // Failure-closed: once tripped, stay tripped. Caller (PR-B2's
      // fork) is expected to abort the active filter session by
      // checking `guard.isTripped()` and forwarding `latency_exceeded`
      // to the JS shim's `wrapTrack` rejection path.
      return false
    }
    if (channelCount !in 1..2) {
      // PR-B2's fork only configures mono or stereo capture; anything
      // else is a programming error worth a fast-fail.
      return false
    }
    lastChannelCount = channelCount

    val startNanos = nowNanos()
    val bytesPerSample = 2 * channelCount
    val frameSamples = samples.remaining() / bytesPerSample
    if (frameSamples == 0) return false
    if (frameSamples > MAX_FRAME_SAMPLES) {
      // Skip rather than reallocate; WebRTC always uses 10ms frames
      // which fit comfortably under MAX_FRAME_SAMPLES.
      return false
    }

    val originalOrder = samples.order()
    samples.order(ByteOrder.LITTLE_ENDIAN)
    val startPos = samples.position()

    // Decode → mono PCM16 in `scratch`.
    if (channelCount == 1) {
      for (i in 0 until frameSamples) {
        scratch[i] = samples.short
      }
    } else {
      for (i in 0 until frameSamples) {
        val l = samples.short.toInt()
        val r = samples.short.toInt()
        scratch[i] = ((l + r) shr 1).toShort()
      }
    }

    // In-place pitch+formant shift via the active shifter
    // (phase vocoder or granular — see USE_PHASE_VOCODER above).
    shifter.process(scratch, 0, scratch, 0, frameSamples, factor)

    // Encode → PCM16 back into the original byte buffer at the
    // original position. Stereo writes the same mono sample to both
    // channels.
    samples.position(startPos)
    if (channelCount == 1) {
      for (i in 0 until frameSamples) {
        samples.putShort(scratch[i])
      }
    } else {
      for (i in 0 until frameSamples) {
        samples.putShort(scratch[i])
        samples.putShort(scratch[i])
      }
    }
    samples.position(startPos)
    samples.order(originalOrder)

    val elapsed = (nowNanos() - startNanos) / 1_000L
    guard.recordFrame(elapsed)
    return true
  }

  /** Reset all DSP state; safe between calls. */
  fun reset() {
    shifter.reset()
    guard.reset()
  }

  /** Diagnostics; do not gate behavior. */
  fun isLatencyTripped(): Boolean = guard.isTripped()

  /** Shift factor (e.g. 0.89 for −2 semitones); diagnostics only. */
  fun shiftFactor(): Float = factor

  companion object {
    /**
     * 60ms at 48kHz mono. WebRTC standard frame is 10ms / 480
     * samples; this gives 6× headroom in case a future ADM bumps to
     * a longer accumulator.
     */
    const val MAX_FRAME_SAMPLES = 2880

    /** Pitch-shifter selection. `true` = phase vocoder (now at a 512
     *  window — see PhaseVocoderPitchShifter), `false` = granular.
     *
     *  1.0.x latency fix. An offline bench of the ACTUAL shifter code
     *  (amplitude-step group-delay + 30s RTF on the JVM, no device)
     *  measured the added call delay + per-frame CPU:
     *    vocoder @1024 (old): 30.3 ms, 253 µs/frame   ← the delay reported
     *    vocoder @512  (now): 16.6 ms, 364 µs/frame   ← shipped
     *    granular:            19.9 ms,   3.6 µs/frame
     *  The vocoder's delay is ~one analysis window, so halving the window
     *  (1024→512) nearly halves the delay — to BELOW granular — while
     *  KEEPING the vocoder's formant control (distinct Smoke/Velvet/Glass,
     *  no granular crackle). CPU is only ~44% over the already-shipped
     *  @1024 vocoder and well under the 10 ms/frame budget even at 15×
     *  mobile slowdown. Granular stays as the `false` fallback (trivial
     *  CPU) if the @512 voice character isn't acceptable on device. */
    private const val USE_PHASE_VOCODER = true
  }
}

/** Common shape so [VoiceFilterDsp] can hold either shifter. */
private interface PitchShifter {
  fun process(
      input: ShortArray,
      inOffset: Int,
      output: ShortArray,
      outOffset: Int,
      n: Int,
      factor: Float,
  )
  fun reset()
}

private class GranularShifterAdapter : PitchShifter {
  private val inner = GranularPitchShifter()
  override fun process(
      input: ShortArray,
      inOffset: Int,
      output: ShortArray,
      outOffset: Int,
      n: Int,
      factor: Float,
  ) = inner.process(input, inOffset, output, outOffset, n, factor)
  override fun reset() = inner.reset()
}

private class PhaseVocoderShifterAdapter(formantFactor: Float = 1f) : PitchShifter {
  private val inner = PhaseVocoderPitchShifter(formantFactor = formantFactor)
  override fun process(
      input: ShortArray,
      inOffset: Int,
      output: ShortArray,
      outOffset: Int,
      n: Int,
      factor: Float,
  ) = inner.process(input, inOffset, output, outOffset, n, factor)
  override fun reset() = inner.reset()
}

File: apps/mobile/ios/SpeakeasyBridges/VoiceFilter/dsp/VoiceFilterDsp.swift
//
//  VoiceFilterDsp.swift
//  Speakeasy — Phase 5j Private Call voice filter
//
//  Swift port of
//  apps/mobile/android/.../voicefilter/dsp/VoiceFilterDsp.kt
//
//  Top-level filter: configurable pitch + formant shift in
//  semitones (default −2; sounds "deeper", more disguised), 120ms
//  p95 latency budget, mono PCM16. Stereo handled by the caller
//  (the audio device collapses stereo→mono BEFORE calling
//  process(...) to avoid duplicate work in the granular shifter).
//
//  Allocation-free per-frame: scratch sized for the largest
//  plausible frame in init().
//

import Foundation

final class VoiceFilterDsp: SampleFilter {

  /// Phase 2a default — flip to `false` to revert to the rc.17
  /// granular shifter if field testing finds the vocoder worse.
  private static let USE_PHASE_VOCODER: Bool = true

  private let factor: Float
  private let formantFactor: Float
  private let shifter: PitchShifter
  private let guardrail: LatencyGuard
  private let nowMicros: () -> Int64

  /// Cached sample-rate so we can detect mid-call reconfigs (e.g.,
  /// route changes) and reset the shifter ring buffer cleanly.
  private var lastSampleRate: Double = 0
  /// Cached channel count for diagnostics; current shifter is mono
  /// only — the audio device collapses stereo→mono upstream.
  private(set) var lastChannelCount: Int = 0

  init(
    semitones: Float = -2.0,
    formantSemitones: Float = 0.0,
    budgetMicros: Int64 = 120_000,
    nowMicros: @escaping () -> Int64 = {
      Int64(DispatchTime.now().uptimeNanoseconds / 1_000)
    }
  ) {
    self.factor = pow(2.0, semitones / 12.0)
    self.formantFactor = pow(2.0, formantSemitones / 12.0)
    // Phase 2a: phase vocoder by default. Lower latency (~10ms vs
    // ~21ms) and no crackle, at the cost of some pitch-shift
    // artifacts on transients and faint metallic edge on sustained
    // vowels. The boolean above lets us flip back to granular if
    // field testing finds the vocoder worse on real hardware.
    // Phase 2b: vocoder gets formantFactor for independent
    // pitch/formant control. Granular ignores it (can't do it).
    self.shifter = Self.USE_PHASE_VOCODER
      ? PhaseVocoderShifterAdapter(formantFactor: self.formantFactor)
      : GranularShifterAdapter()
    self.guardrail = LatencyGuard(budgetMicros: budgetMicros)
    self.nowMicros = nowMicros
  }

  func process(
    samples: UnsafeMutablePointer<Int16>,
    frameCount: Int,
    channelCount: Int,
    sampleRateHz: Double
  ) -> Bool {
    if guardrail.isTripped() { return false }
    guard channelCount == 1 || channelCount == 2 else { return false }
    if frameCount <= 0 { return false }
    if frameCount > Self.maxFrameSamples { return false }

    lastChannelCount = channelCount
    if lastSampleRate != sampleRateHz {
      lastSampleRate = sampleRateHz
      shifter.reset()
    }

    let start = nowMicros()
    if channelCount == 1 {
      shifter.process(input: samples, output: samples, count: frameCount, factor: factor)
    } else {
      // Stereo path: collapse to mono into a scratch, shift, then
      // duplicate mono back to L and R. The audio device collapses
      // BEFORE calling us when it can; this branch is a safety net.
      var mono = [Int16](repeating: 0, count: frameCount)
      for i in 0..<frameCount {
        let l = Int32(samples[i * 2])
        let r = Int32(samples[i * 2 + 1])
        mono[i] = Int16((l + r) >> 1)
      }
      mono.withUnsafeMutableBufferPointer { buf in
        guard let base = buf.baseAddress else { return }
        shifter.process(input: base, output: base, count: frameCount, factor: factor)
      }
      for i in 0..<frameCount {
        samples[i * 2] = mono[i]
        samples[i * 2 + 1] = mono[i]
      }
    }
    let elapsed = nowMicros() - start
    guardrail.recordFrame(elapsedMicros: elapsed)
    return true
  }

  /// Diagnostics; do not gate behavior. The audio device queries
  /// this after a `process(...)=false` to decide whether to emit
  /// `latency_exceeded` to the JS shim's wrapTrack rejection.
  func isLatencyTripped() -> Bool { guardrail.isTripped() }

  func shiftFactor() -> Float { factor }

  func reset() {
    shifter.reset()
    guardrail.reset()
    lastSampleRate = 0
  }

  /// 60ms at 48kHz mono — same headroom as the Android DSP.
  /// WebRTC standard frame is 10ms (480 samples at 48kHz); 6×
  /// margin keeps us safe if a future RTCAudioDevice configuration
  /// bumps to a longer IO buffer.
  static let maxFrameSamples = 2880
}

/// Common shape so [VoiceFilterDsp] can hold either shifter.
protocol PitchShifter {
  func process(
    input: UnsafePointer<Int16>,
    output: UnsafeMutablePointer<Int16>,
    count: Int,
    factor: Float
  )
  func reset()
}

final class GranularShifterAdapter: PitchShifter {
  private let inner = GranularPitchShifter()
  func process(
    input: UnsafePointer<Int16>,
    output: UnsafeMutablePointer<Int16>,
    count: Int,
    factor: Float
  ) {
    inner.process(input: input, output: output, count: count, factor: factor)
  }
  func reset() { inner.reset() }
}

final class PhaseVocoderShifterAdapter: PitchShifter {
  private let inner: PhaseVocoderPitchShifter
  init(formantFactor: Float = 1.0) {
    self.inner = PhaseVocoderPitchShifter(formantFactor: formantFactor)
  }
  func process(
    input: UnsafePointer<Int16>,
    output: UnsafeMutablePointer<Int16>,
    count: Int,
    factor: Float
  ) {
    inner.process(input: input, output: output, count: count, factor: factor)
  }
  func reset() { inner.reset() }
}
